### PR TITLE
[Coop] Cleanup ahead of converting ves_icall_System_Threading_Thread_…

### PR DESCRIPTION
**Pull request by jaykrell**
…GetAbortExceptionState.

 - serialize_object and deserialize_object are almost identical and
   can be implemented on top of the same function.
 - Error propagation was a little wierd and had FIXME comments.
   It involved MonoError, MonoException, and gboolean *failed, and
   some conversions. Converge on mostly MonoError, with fewer conversions,
   albeit still sometimes holding a MonoException.

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build failed

File: mono/metadata/object.c [hunk 1]
```diff
@@ -4395,76 +4395,43 @@ mono_runtime_try_run_main (MonoMethod *method, int argc, char* argv[],
 	return mono_runtime_try_exec_main (method, args, exc);
 }
 
-
 static MonoObject*
-serialize_object (MonoObject *obj, gboolean *failure, MonoObject **exc)
+serialize_or_deserialize_object (MonoObject *obj, MonoError *error,
+	const gchar *method_name, MonoMethod **method)
 {
```
Comment (vargaz) [inline mono/metadata/object.c]: error is usually passed as the last argument.

File: mono/metadata/object.c [hunk 2]
```diff
@@ -4395,76 +4395,43 @@ mono_runtime_try_run_main (MonoMethod *method, int argc, char* argv[],
...
 		MonoClass *klass = mono_class_get_remoting_services_class ();
-		serialize_method = mono_class_get_method_from_name (klass, "SerializeCallData", -1);
+		*method = mono_class_get_method_from_name (klass, method_name, -1);
 	}
 
-	if (!serialize_method) {
-		*failure = TRUE;
+	if (!*method) {
+		mono_error_set_exception_instance (error, NULL);
 		return NULL;
```
Comment (vargaz) [inline mono/metadata/object.c]: What will this do with a NULL argument ? I think this could be just an assert.

File: mono/metadata/object.c [hunk 3]
```diff
@@ -4395,76 +4395,43 @@ mono_runtime_try_run_main (MonoMethod *method, int argc, char* argv[],
...
 	}
 
-	g_assert (!mono_class_is_marshalbyref (mono_object_class (obj)));
-
```
Comment (vargaz) [inline mono/metadata/object.c]: This could stay.

File: mono/metadata/object.c (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/object.c]: I was kinda appending parameters, albeit having replaced failure/exc with error; ok.

File: mono/metadata/object.c (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/object.c]: I'm emulating the previous behavior I believe -- where it'd set *failure = TRUE and leave exc == NULL and caller would mono_error_set_exception_instance.

File: mono/metadata/object.c [hunk 4]
```diff
@@ -4395,76 +4395,43 @@ mono_runtime_try_run_main (MonoMethod *method, int argc, char* argv[],
...
-	array = mono_runtime_try_invoke (serialize_method, NULL, params, exc, error);
-	if (*exc == NULL && !mono_error_ok (error))
-		*exc = (MonoObject*) mono_error_convert_to_exception (error); /* FIXME convert serialize_object to MonoError */
-	else
-		mono_error_cleanup (error);
+	MonoException *exc = NULL;
+	void *params [ ] = { obj };
+	MonoObject *result = mono_runtime_try_invoke (*method, NULL, params, (MonoObject**)&exc, error);
+	if (!is_ok (error) && exc)
```
Comment (jaykrell) [inline mono/metadata/object.c]: This is off -- remove the exclamation point.

File: mono/metadata/object.c (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/object.c]: I think "it works" -- the error handling infrastructure will take it -- but maybe indeed assert is preferred.

**Comment by jaykrell** [comment]
@monojenkins build  Linux ARMv7 Interpreter

**Comment by jaykrell** [comment]
@monojenkins build Linux ARMv7 Interpreter

**Pull request merged**